features/285-design-webhooks/events.md

```diff
@@ -0,0 +1,259 @@
+# Web-hook Events
+
+All web-hook events have the same format:
+```json
+{
+    "id": "0E1E9537-DF4F-4AF6-8B48-3DB4574D4F24",
+    "seq": 435,
+    "type": "after_user_create",
+    "payload": {
+        "version": 1
+        // ...
+    },
+    "context": {
+        // ...
+    }
+}
+```
+
+- `id`: the ID of the web-hook event.
+- `seq`: monotonic increasing event sequence number to establish event order.
+- `type`: the type of the web-hook event.
+- `payload`: the payload of the web-hook; it is specific to each event type.
+- `context`: the context of the web-hook.
+- `version`: the version of the web-hook payload; it is specific to the event
+             type.
+
+## Versioning
+The content of top-level fields `type` and `context` are guaranteed to exist
+and only backward-compatible changes would be made.
+
+The content of top-level field `payload` are specific to each event type.
+Field `version` in the payload would be updated when a backward incompatible
+change is made. Consumer should handle unknown version gracefully.
+
+## Context
+Each web-hook event would have a context, it contains the environmental
+information about the event:
+
+- `timestamp`:
+  timestamp of when this event is generated; retried deliveries would not
+  affect this field.
+- `request_id`:
+  ID of the HTTP request that generated this event; if the event is not
+  generated by HTTP request, this field would be `null`.
+- `user_id`:
+  the ID of user performing the request that generated this event; if request
+  is not authenticated or event is not generated by user action, this field
+  would be `null`.
+- `principal_id`:
+  the ID of principal performing the request that generated this event; if
+  request is not authenticated or event not generated by user action, this field
+  would be `null`.
+
+**NOTE**
+- When user is performing authentication (e.g. signing up/logging in),
+  the request will not be treated as authenticated with the user to
+  sign up/log in, in both BEFORE/AFTER events.
+
+## Event Types
+
+### before_user_create, after_user_create
+```json
+{
+    "user": { /* a User object */ },
+    "identities": [
+        { /* an Identity object */ },
+        { /* an Identity object */ }
+    ]
+}
+```
+
+### before_identity_create, after_identity_create
+```json
+{
+    "is_user_creating": false,
```

Comment:
I think it's better to refer to the specific operation (user creation) in the flag's name.
---
I intended this flag to be used to distinguish whether the event is triggered along with user_create events; developer should do user create specific actions in user_create handlers, and may want to avoid duplicate works in identity_create handler.

Most of the time, the handler would only need to user ID from the event, which is present in the identity object. I don't think the user object is needed in the event.